src/v/cluster/decommissioning_monitor.cc

```diff
@@ -0,0 +1,256 @@
+  ss::sharded<cluster::topics_frontend>& fr,
+  ss::sharded<topic_table>& tt,
+  ss::sharded<partition_allocator>& pal,
+  consensus_ptr raft0,
+  ss::sharded<ss::abort_source>& as)
+  : _topics_frontend(fr)
+  , _topics(tt)
+  , _allocator(pal)
+  , _raft0(raft0)
+  , _as(as)
+  , _retry_timeout(config::shard_local_cfg().decomission_monitor_retry_ms()) {
+    _retry_timer.set_callback([this] {
+        // just signal new changes
+        _new_changes.signal();
+    });
+}
+
+ss::future<> decommissioning_monitor::stop() {
+    _retry_timer.cancel();
+    _new_changes.broken();
+    return _bg.close();
+}
+
+ss::future<> decommissioning_monitor::start() {
+    start_decommission_loop();
+    return ss::now();
+}
+
+void decommissioning_monitor::start_decommission_loop() {
+    return (void)ss::with_gate(_bg, [this] {
+        return ss::do_until(
+          [this] { return _as.local().abort_requested(); },
+          [this] {
+              return _new_changes.wait()
+                .then([this] {
+                    return ss::parallel_for_each(
+                      _reallocations, [this](reallocation_meta& meta) {
+                          return reallocate_replica_set(meta);
+                      });
+                })
+                .then([this] {
+                    // remove finished meta
+                    auto it = std::remove_if(
+                      std::begin(_reallocations),
+                      std::end(_reallocations),
+                      [](reallocation_meta& meta) { return meta.is_finished; });
+
+                    _reallocations.erase(it, std::end(_reallocations));
+                })
+                .then([this] {
+                    return _allocator.invoke_on(
+                      partition_allocator::shard,
+                      [](partition_allocator& allocator) {
+                          std::vector<model::node_id> ids;
+                          for (const auto& [id, node] :
+                               allocator.allocation_nodes()) {
+                              if (node->is_decommissioned() && node->empty()) {
+                                  ids.push_back(id);
+                              }
+                          }
+                          return ids;
+                      });
+                })
+                .then([this](std::vector<model::node_id> to_remove) {
+                    // no nodes to reamove from cluster
+                    if (to_remove.empty() || !_raft0->is_leader()) {
+                        return ss::now();
+                    }
+                    vlog(
+                      clusterlog.info,
+                      "removing decommissioned nodes: {}",
+                      to_remove);
+                    // we are not interested in the result as we check
+                    // configuration in next step
+                    return _raft0
+                      ->remove_members(to_remove, model::revision_id{0})
```

Comment:
why a fixed revision?

Reply:
currently  in raft 0 we use fixed revision, it is because we do not yet assign unique identities to nodes when then join the cluster, there is a future plan to solve this.